src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -149,15 +185,12 @@ public StaticHostProvider(Collection<InetSocketAddress> serverAddresses,
      * @param currentHost the host to which this client is currently connected
      * @return true if changing connections is necessary for load-balancing, false otherwise  
      */
-
-
     @Override
     public synchronized boolean updateServerList(
             Collection<InetSocketAddress> serverAddresses,
             InetSocketAddress currentHost) {
-        // Resolve server addresses and shuffle them
-        List<InetSocketAddress> resolvedList = resolveAndShuffle(serverAddresses);
-        if (resolvedList.isEmpty()) {
+        List<InetSocketAddress> shuffledList = shuffle(serverAddresses);
```

Comment:
I am not sure why we change the `resolveAndShuffle` to `shuffle` here, they are semantically different (one tries to resolve address, the other does not and only does shuffle.). The `serverAddresses` passed in this method is unresolved address, and we need it resolved because we rely on the resolved addresses to compare the old / new server list (in the context of probability rebalancing clients for dynamic reconfiguration). 

Without resolving I think the client rebalance logic will be broken. A side note that all tests still passed probably indicate we don't have a 100% coverage for the logic in our tests.

Reply:
After this change, the old / new server list both contains unresolved address, so it's comparable, but it's meaningless to compare addr.getAddress because it will return null when it's unresolved.

Reply:
`resolveAndShuffle` has been split into two methods: `resolve` and `shuffle`, because we have to do it separately. Github shows the diff as original has been renamed and new one has been created.

I'll double check the rebalance functionality to make sure it's working properly and will also check related unit tests. I think what @lvfangmin is saying makes sense.

Reply:
Comparison works for unresolved addresses too, because of the last if condition here:
```java
if (addr.getPort() == myServer.getPort()
    && ((addr.getAddress() != null
    && myServer.getAddress() != null && addr
    .getAddress().equals(myServer.getAddress())) || addr
    .getHostString().equals(myServer.getHostString()))) {
```
As long as `getHostString()` works for unresolved addresses, we're fine. However, exactly the same functionality has already been implemented in the `InetSocketAddress.equals()` method, so  I refactored this part to use it.

Reply:
>> However, exactly the same functionality has already been implemented in the InetSocketAddress.equals() method

I am wondering if this is the case or not. I just did a random peek at https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/net/InetSocketAddress.java#L300, looks like if we compare an unresolved address with a resolved address, the equals will return false - but the code you pasted will return true if getHostString works for both resolved and unresolved address... could you double check this behavior?

Also, is it possible to add a test case to cover the case where the second parameter of updateServerList is a resolved address? The existing test cases only cover the case where the second parameter (myServer) is unresolved. In practice I think the method updateServerList is called by ZooKeeper's updateServerList method with second parameter as a resolved address (the remote server where current client connected to.).

Reply:
@hanm Good catch, that makes perfect sense. The implementation is slightly different.

I added 4 new test cases to cover the different combinations of the 2nd and 3rd parameter. Some of them could be redundant (as you said tests present already), but I've already opened a Jira to clean-up this test file, so it should be okay. Test scenarios:

*Given:* list of servers contains 1 element, client is connected (currentHost (myServer) != null), trying to replace server list with the same address (replaceHost).

1. `currentHost` resolved, `replaceHost` unresolved => client should disconnect,
2. `currentHost` resolved, `replaceHost` resolved => client should *not* disconnect,
3. `currentHost` unresolved, `replaceHost` unresolved => client should disconnect,
4. `currentHost` unresolved, `replaceHost` resolved => client should disconnect

My implementation covers all of these scenarios. The only difference is in *case no.4* where the original impl returns false (client should not disconnect). Question is what we would like to do in the 4th case? Does it make sense at all, as you mentioned client should not be connected to an unresolved address?

Reply:
Thanks for quick update, @anmolnar ! 

>> The built-in implementation treats them as different address

That's what I am not sure about. If an unresolved address and a resolved address actually maps same address (after unresolved address gets resolved), should both be treated as same address (old behavior, via getHostString), instead of different address? 

What I am thinking is to instead of using equal - use the old verbose code which will work (so far) for all combinations of comparing resolved and unresolved address. What do you think about this?

Reply:
My only concern is with the first case: IP address might have changed of the host, so client should disconnect and trigger a re-resolution.

Anyway. No issue has been reported with the logic we're talking about, so let's just leave it out from this patch. Reverted.